src/loop/parser.ml

```diff
@@ -190,70 +192,131 @@ module Make(State : State.S) = struct
     (* Parse the input *)
-    let st, file, g =
-      try
-        match file.source with
-        | `Stdin ->
-          let lang, file_loc, gen, _ = Logic.parse_input
-              ?language:file.lang (`Stdin (Logic.Smtlib2 `Latest))
-          in
-          let file = { file with loc = file_loc; lang = Some lang; } in
-          st, file, gen
-        | `Raw (filename, contents) ->
-          let lang =
-            match file.lang with
-            | Some l -> l
-            | None ->
-              let res, _, _ = Logic.of_filename filename in
-              res
-          in
-          begin match file.mode with
-          | None
-          | Some `Incremental ->
-            let lang, file_loc, gen, cl = Logic.parse_input
-                ~language:lang (`Raw (filename, lang, contents)) in
-            let file = { file with loc = file_loc; lang = Some lang; } in
-            st, file, gen_finally gen cl
-          | Some `Full ->
-            let lang, file_loc, l = Logic.parse_raw_lazy ~language:lang
-              ~filename contents
-            in
-            let file = { file with loc = file_loc; lang = Some lang; } in
-            st, file, gen_of_llist l
-          end
-        | `File f ->
-          let s = Dolmen.Std.Statement.include_ f [] in
-          (* Auto-detect input format *)
-          let lang =
-            match file.lang with
-            | Some l -> l
-            | None ->
-              let res, _, _ = Logic.of_filename f in
-              res
-          in
-          (* Formats Dimacs and Tptp are descriptive and lack the emission
-              of formal solve/prove instructions, so we need to add them. *)
-          let s' =
-            match lang with
-            | Logic.Zf
-            | Logic.ICNF
-            | Logic.Smtlib2 _
-            | Logic.Alt_ergo -> s
-            | Logic.Dimacs
-            | Logic.Tptp _ ->
-              Dolmen.Std.Statement.pack [s; Dolmen.Std.Statement.prove ()]
-          in
-          let file = { file with lang = Some lang; } in
-          st, file, (Gen.singleton s')
-      with
-      | Logic.Extension_not_found ext ->
-        State.error st extension_not_found ext, file, Gen.empty
+    match source with
+    | `Raw (filename, contents) ->
+      begin match file.mode with
+      | None
+      | Some `Incremental ->
+        let lang, file_loc, gen, cl = Logic.parse_input
+            ~language:lang (`Raw (filename, lang, contents)) in
+        let file = { file with loc = file_loc; lang = Some lang; } in
+        st, file, gen_finally gen cl
+      | Some `Full ->
+        let lang, file_loc, l = Logic.parse_raw_lazy ~language:lang
+          ~filename contents
+        in
+        let file = { file with loc = file_loc; lang = Some lang; } in
+        (* We are called from inside a [gen'] so the returned generator is
+        always evaluated immediately, there is no need to add laziness *)
+        st, file, Gen.of_list (Lazy.force l)
```

Comment:
That's not good: the lazy part was actually added so that any lexer/parser exception (raise in case of errors during lexing/parsing), are actually raised by the generator, instead of during the `parse_*` functions (this is actually important because the generator is wrapped so that exceptions are properly watched and transformed into errors acting on the most up-to-date state).